data-structures-csharp/data-structures-csharp/CircularBuffer/CircularBuffer.cs

```diff
@@ -32,52 +33,81 @@ private void ObjectInvariant()
             Contract.Invariant(Count <= Capacity);
         }
 
-        public CircularBuffer() 
-        {
-            internalList = new List<T>();
-        }
-
+        /// <summary>
+        /// Instantiate a new circular buffer.
+        /// </summary>
+        /// <param name="capacity">The maximum size of the buffer before it begins overwriting itself.</param>
         public CircularBuffer(int capacity)
         {
             Contract.Requires<ArgumentOutOfRangeException>(capacity > 0);
-
-            internalList = new List<T>(capacity);
+            buffer = new T[capacity];
+            Capacity = capacity;
+            Count = 0;
+            addIndex = 0;
+            getIndex = 0;
         }
 
+        /// <summary>
+        /// Add an element to the circular buffer.
+        /// </summary>
         public void Add(T item)
         {
-            throw new NotImplementedException();
+            Contract.Requires<ArgumentNullException>(item != null);
+            buffer[addIndex] = item;
+
+            if (Count < Capacity)
+                Count++;
+
+            if (addIndex == getIndex)
+                getIndex = getIndex < Capacity - 1 ? getIndex + 1 : 0;
+
+            addIndex = addIndex < Capacity - 1 ? addIndex + 1 : 0;
         }
 
+        /// <summary>
+        /// Clears the circular buffer of all elements.
+        /// </summary>
         public void Clear()
         {
-            internalList.Clear();
+            buffer = new T[Capacity];
+            Count = 0;
+            addIndex = 0;
+            getIndex = 0;
         }
 
+        /// <summary>
+        /// Determines whether a sequence contains a specified element by using the default equality comparer.
+        /// </summary>
         public bool Contains(T item)
         {
             Contract.Requires<ArgumentNullException>(item != null);
-            return internalList.Contains(item);
-        }
-
-        public void CopyTo(T[] array, int arrayIndex)
-        {
-            throw new NotImplementedException();
+            return buffer.Contains(item);
         }
 
-        public bool IsReadOnly
+        /// <summary>
+        /// Copy the circular buffer to an array starting at the specified index in the destination array.
+        /// </summary>
+        public void CopyTo(T[] array, int index = 0)
         {
-            get { throw new NotImplementedException(); }
+            Contract.Requires<ArgumentNullException>(array != null);
```

Comment:
Add preconditions if array can hold the elements i.e., "Contract.Requires<ArgumentNullException>((index-Capacity+1)<=array.length)"